### PR TITLE
clean up cluster resources after failure

### PR DESCRIPTION
**Pull request by moritzjohner-form3**
This PR:
1. bumps k8s dependencies 0.22 -> 0.24
2. bump go 1.17 to go 1.19
3. adds small feature to clean up resources after cluster failure

⚠️ This requires an update to RBAC clusterrole, see PR ⚠️ 

### Proof of work
kind cluster with:
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: control-plane
- role: control-plane
```

etcd cluster with:
```yaml
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdCluster"
metadata:
  name: "example-etcd-cluster"
  ## Adding this annotation make this cluster managed by clusterwide operators
  ## namespaced operators ignore it
  # annotations:
  #   etcd.database.coreos.com/scope: clusterwide
spec:
  pod:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
            - matchExpressions:
                - key: node-role.kubernetes.io/control-plane
                  operator: In
                  values:
                    - ""
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - topologyKey: kubernetes.io/hostname
            labelSelector:
              matchLabels:
                etcd_cluster: cilium-etcd
    etcdEnv:
      # enable periodic compaction to avoid performance degradation and eventual storage space exhaustion
      # https://etcd.io/docs/v3.4/op-guide/maintenance/#history-compaction
      - name: ETCD_AUTO_COMPACTION_RETENTION
        value: 1h
    resources:
      requests:
        cpu: 100m
        memory: 200Mi
    labels:
      io.cilium/app: etcd-operator
    tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
        operator: Equal
        value: ""
      - effect: NoSchedule
        key: node-role.kubernetes.io/control-plane
        operator: Equal
        value: ""
  failurePolicy: Recreate
  scalingTimeout: 3m
  size: 3
  version: "3.4.7"
```

1. apply etcd resource
2. kill the last of the three pods until the clusters gets recreated
4. verify that old pods/services get cleaned up

Logs:
```
time="2023-03-29T12:52:14Z" level=info msg="cluster membership: example-etcd-cluster-z22d6n9rts,example-etcd-cluster-4np9wvtsfx,example-etcd-cluster-rcjrp7zjzh" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:14Z" level=info msg="removing one dead member" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:14Z" level=info msg="removing dead member \"example-etcd-cluster-rcjrp7zjzh\"" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:14Z" level=info msg="removed member (example-etcd-cluster-rcjrp7zjzh) with ID (12994523927688839123)" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:14Z" level=info msg="Finish reconciling" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
... 
time="2023-03-29T12:52:22Z" level=info msg="scaling timeout reached, cluster creation failed" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:22Z" level=info msg="cluster failed. Reporting failed reason..." cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:22Z" level=info msg="deleting cluster due to failurePolicy=Recreate" pkg=controller
time="2023-03-29T12:52:22Z" level=info msg="cluster is deleted by user" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:22Z" level=info msg="cleaning up cluster resources" pkg=controller
time="2023-03-29T12:52:22Z" level=info msg="creating cluster with Spec:" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:22Z" level=info msg="{" cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:22Z" level=info msg="    \"size\": 3," cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
time="2023-03-29T12:52:22Z" level=info msg="    \"repository\": \"quay.io/coreos/etcd\"," cluster-name=example-etcd-cluster cluster-namespace=default pkg=cluster
...
```

### CODE REVIEW

File: cmd/operator/main.go [hunk 1]
```diff
@@ -109,7 +109,7 @@ func main() {
 	http.Handle("/metrics", promhttp.Handler())
 	go http.ListenAndServe(listenAddr, nil)
 
-	rl, err := resourcelock.New(resourcelock.EndpointsResourceLock,
+	rl, err := resourcelock.New(resourcelock.EndpointsLeasesResourceLock,
```
Comment (moritzjohner-form3) [inline cmd/operator/main.go:112]: This change is required, b/c `EndpointsResourceLock` has been removed. ⚠️ This requires RBAC changes.

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -90,7 +90,7 @@ type Cluster struct {
 
 func New(config Config, cl *api.EtcdCluster) *Cluster {
 	lg := logrus.WithField("pkg", "cluster").WithField("cluster-name", cl.Name).WithField("cluster-namespace", cl.Namespace)
-	if len(cl.Name) > k8sutil.MaxNameLength || len(cl.ClusterName) > k8sutil.MaxNameLength {
```
Comment (moritzjohner-form3) [inline pkg/cluster/cluster.go:93]: `cl.ClusterName` was part of `metav1` and has been deprecated.

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -90,7 +90,7 @@ type Cluster struct {
...
+	if len(cl.Name) > k8sutil.MaxNameLength || len(cl.GetZZZ_DeprecatedClusterName()) > k8sutil.MaxNameLength {
```
Comment (moritzjohner-form3) [inline pkg/cluster/cluster.go:93]: `cl.ClusterName` was part of `metav1` and has been deprecated.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -84,9 +89,18 @@ func (c *Controller) handleClusterEvent(event *Event) (bool, error) {
...
 
+			// operator does not clean up resources on it's own, it relies
+			// on ownerReferences / k8s gc to cleanup orphaned pods
+			// for this case we want to clean up manually
+			c.logger.Info("cleaning up cluster resources")
+			err := c.cleanupClusterResources(clus)
+			if err != nil {
+				c.logger.Errorf("unable to cleanup cluster resources: %v", err)
+			}
+
```
Comment (moritzjohner-form3) [inline pkg/controller/controller.go:100]: This is the logical change: cleans up pods, peer-service and client-service of this particular cluster.

**Review by stevo-f3 — APPROVED** [review]

**Pull request merged**